osu.Game/Rulesets/RulesetStore.cs

```diff
@@ -96,13 +96,25 @@ private void addMissingRulesets()
 
                 context.SaveChanges();
 
-                // add any other modes
                 var existingRulesets = context.RulesetInfo.ToList();
 
+                // add any other rulesets which have assemblies present but are not yet in the database.
                 foreach (var r in instances.Where(r => !(r is ILegacyRuleset)))
                 {
                     if (existingRulesets.FirstOrDefault(ri => ri.InstantiationInfo.Equals(r.RulesetInfo.InstantiationInfo, StringComparison.Ordinal)) == null)
-                        context.RulesetInfo.Add(r.RulesetInfo);
+                    {
+                        var existingSameShortName = existingRulesets.FirstOrDefault(ri => ri.ShortName == r.RulesetInfo.ShortName);
+
+                        if (existingSameShortName != null)
+                        {
+                            // even if a matching InstantiationInfo was not found, there may be an existing ruleset with the same ShortName.
+                            // this generally means the user or ruleset provider has renamed their dll but the underlying ruleset is *likely* the same one.
+                            // in such cases, update the instantiation info of the existing entry to point to the new one.
+                            existingSameShortName.InstantiationInfo = r.RulesetInfo.InstantiationInfo;
```

Comment:
Does this actually commit the updated InstantiationInfo to the database? It doesn't seem like it would, if I'm understanding the code correctly.

Reply:
It did seem to, I'll double check. Which part looks like it wouldn't?

Reply:
That `existingRulesets` is a local list of plain-old-data objects, should be completely invisible to EF (it wouldn't be the case if this were Realm, I think?).

Reply:
It works as expected.

EF is smarter than you may expect. It tracks references internally.